File: frontend/src/services/comptabilite.ts
import { api } from './api';
import { format } from 'date-fns';
import type { ApiResponse } from '../types/common';
import {
  CompteComptable,
  EcritureComptable,
  JournalComptable,
  GrandLivre,
  CompteBalance,
  ComptabiliteStats,
  CompteComptableFormData,
  EcritureComptableFormData,
  TypeCompteComptable,
  StatutEcriture,
  BudgetAnalysis,
  Balance,
  RapportComptable
} from '../types/comptabilite';

const API_BASE = '/api/v1/comptabilite';

// Statistiques et analyses
export const getComptabiliteStats = async (): Promise<ComptabiliteStats> => {
  const response = await api.get<ApiResponse<ComptabiliteStats>>(`${API_BASE}/stats`);
  return response.data.data;
};

export const getBudgetAnalysis = async (periode: string): Promise<BudgetAnalysis> => {
  const response = await api.get<ApiResponse<BudgetAnalysis>>(`${API_BASE}/budget/analysis`, {
    params: { periode }
  });
  return response.data.data;
};

export const getCashFlow = async (days: number = 30) => {
  const response = await api.get<ApiResponse<any>>(`${API_BASE}/cashflow`, {
    params: { days }
  });
  return response.data.data;
};

// Comptes comptables
export const getComptes = async (params?: {
  type_compte?: TypeCompteComptable;
  actif?: boolean;
}): Promise<CompteComptable[]> => {
  const response = await api.get<ApiResponse<CompteComptable[]>>(`${API_BASE}/comptes`, { params });
  return response.data.data;
};

export const getCompte = async (id: string): Promise<CompteComptable> => {
  const response = await api.get<ApiResponse<CompteComptable>>(`${API_BASE}/comptes/${id}`);
  return response.data.data;
};

export const createCompte = async (data: CompteComptableFormData): Promise<CompteComptable> => {
  const response = await api.post<ApiResponse<CompteComptable>>(`${API_BASE}/comptes`, data);
  return response.data.data;
};

export const updateCompte = async (id: string, data: Partial<CompteComptableFormData>): Promise<CompteComptable> => {
  const response = await api.patch<ApiResponse<CompteComptable>>(`${API_BASE}/comptes/${id}`, data);
  return response.data.data;
};

// Écritures comptables
export const getEcritures = async (params?: {
  compte_id?: string;
  journal_id?: string;
  date_debut?: Date;
  date_fin?: Date;
  statut?: StatutEcriture;
}): Promise<EcritureComptable[]> => {
  const formattedParams = {
    ...params,
    date_debut: params?.date_debut ? format(params.date_debut, 'yyyy-MM-dd') : undefined,
    date_fin: params?.date_fin ? format(params.date_fin, 'yyyy-MM-dd') : undefined,
  };
  const response = await api.get<ApiResponse<EcritureComptable[]>>(`${API_BASE}/ecritures`, { params: formattedParams });
  return response.data.data;
};

export const createEcriture = async (data: EcritureComptableFormData): Promise<EcritureComptable> => {
  const response = await api.post<ApiResponse<EcritureComptable>>(`${API_BASE}/ecritures`, data);
  return response.data.data;
};

export const validerEcriture = async (id: string, validee_par_id: string): Promise<EcritureComptable> => {
  const response = await api.post<ApiResponse<EcritureComptable>>(`${API_BASE}/ecritures/${id}/valider`, {
    validee_par_id
  });
  return response.data.data;
};

// Journaux comptables
export const getJournaux = async (params?: {
  type?: string;
  actif?: boolean;
}): Promise<JournalComptable[]> => {
  const response = await api.get<ApiResponse<JournalComptable[]>>(`${API_BASE}/journaux`, { params });
  return response.data.data;
};

export const getJournal = async (id: string): Promise<JournalComptable> => {
  const response = await api.get<ApiResponse<JournalComptable>>(`${API_BASE}/journaux/${id}`);
  return response.data.data;
};

export const createJournal = async (data: Partial<JournalComptable>): Promise<JournalComptable> => {
  const response = await api.post<ApiResponse<JournalComptable>>(`${API_BASE}/journaux`, data);
  return response.data.data;
};

// Rapports
export const getGrandLivre = async (params?: {
  compte_id?: string;
  date_debut?: Date;
  date_fin?: Date;
}): Promise<GrandLivre[]> => {
  const formattedParams = {
    ...params,
    date_debut: params?.date_debut ? format(params.date_debut, 'yyyy-MM-dd') : undefined,
    date_fin: params?.date_fin ? format(params.date_fin, 'yyyy-MM-dd') : undefined,
  };
  const response = await api.get<ApiResponse<GrandLivre[]>>(`${API_BASE}/grand-livre`, { params: formattedParams });
  return response.data.data;
};

export const getBalance = async (params?: {
  date_debut?: Date;
  date_fin?: Date;
  type_compte?: TypeCompteComptable;
}): Promise<CompteBalance[]> => {
  const formattedParams = {
    ...params,
    date_debut: params?.date_debut ? format(params.date_debut, 'yyyy-MM-dd') : undefined,
    date_fin: params?.date_fin ? format(params.date_fin, 'yyyy-MM-dd') : undefined,
  };
  const response = await api.get<ApiResponse<CompteBalance[]>>(`${API_BASE}/balance`, { params: formattedParams });
  return response.data.data;
};

export const getBilan = async (date_fin: Date): Promise<Balance> => {
  const response = await api.get<ApiResponse<Balance>>(`${API_BASE}/bilan`, {
    params: { date_fin: format(date_fin, 'yyyy-MM-dd') }
  });
  return response.data.data;
};

export const getCompteResultat = async (date_debut: Date, date_fin: Date): Promise<RapportComptable> => {
  const response = await api.get<ApiResponse<RapportComptable>>(`${API_BASE}/compte-resultat`, {
    params: {
      date_debut: format(date_debut, 'yyyy-MM-dd'),
      date_fin: format(date_fin, 'yyyy-MM-dd')
    }
  });
  return response.data.data;
};

// Export
export const exporterDonnees = async (params: { type: string; dateDebut: string; dateFin: string; format: string }) => {
  const queryParams = new URLSearchParams({
    type: params.type,
    date_debut: params.dateDebut,
    date_fin: params.dateFin,
    format: params.format
  });
  const response = await api.get<Blob>(`${API_BASE}/export?${queryParams}`, {
    responseType: 'blob'
  });
  return response.data;
};

File: api/v1/endpoints/comptabilite.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import List, Optional, Dict, Any
from datetime import date
from uuid import UUID

from db.database import get_db
from services.comptabilite_service import ComptabiliteService
from schemas.comptabilite import (
    CompteComptableCreate, CompteComptableUpdate, CompteComptableResponse,
    EcritureComptableCreate, EcritureComptableUpdate, EcritureComptableResponse,
    JournalComptableCreate, JournalComptableResponse,
    ExerciceComptableCreate, ExerciceComptableResponse,
    LigneGrandLivre, CompteBalance, BilanResponse, CompteResultatResponse
)

router = APIRouter(prefix="/comptabilite", tags=["comptabilite"])

# [Garder tous les endpoints existants jusqu'à get_compte_resultat]

# Nouveaux endpoints pour les statistiques et analyses
@router.get("/stats")
async def get_stats(
    db: Session = Depends(get_db)
) -> Dict[str, Any]:
    """Récupère les statistiques financières"""
    service = ComptabiliteService(db)
    try:
        return await service.get_stats()
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/budget/analysis")
async def get_budget_analysis(
    periode: str = Query(..., description="Période d'analyse (format: YYYY-MM)"),
    db: Session = Depends(get_db)
) -> Dict[str, Any]:
    """Analyse détaillée du budget pour une période"""
    service = ComptabiliteService(db)
    try:
        return await service.get_budget_analysis(periode)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/cashflow")
async def get_cashflow(
    days: int = Query(30, description="Nombre de jours d'historique", ge=1, le=365),
    db: Session = Depends(get_db)
) -> List[Dict[str, Any]]:
    """Récupère les données de trésorerie sur une période"""
    service = ComptabiliteService(db)
    try:
        return await service.get_cashflow(days)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

# Endpoints pour les comptes comptables
@router.post("/comptes", response_model=CompteComptableResponse)
async def create_compte(
    compte: CompteComptableCreate,
    db: Session = Depends(get_db)
):
    """Crée un nouveau compte comptable"""
    service = ComptabiliteService(db)
    return await service.create_compte(compte.dict())

@router.get("/comptes", response_model=List[CompteComptableResponse])
async def list_comptes(
    type_compte: Optional[str] = None,
    actif: Optional[bool] = None,
    db: Session = Depends(get_db)
):
    """Liste les comptes comptables avec filtres optionnels"""
    service = ComptabiliteService(db)
    return await service.get_comptes(type_compte=type_compte, actif=actif)

@router.get("/comptes/{compte_id}", response_model=CompteComptableResponse)
async def get_compte(
    compte_id: UUID,
    db: Session = Depends(get_db)
):
    """Récupère les détails d'un compte comptable"""
    service = ComptabiliteService(db)
    compte = await service.get_compte(compte_id)
    if not compte:
        raise HTTPException(status_code=404, detail="Compte non trouvé")
    return compte

@router.patch("/comptes/{compte_id}", response_model=CompteComptableResponse)
async def update_compte(
    compte_id: UUID,
    compte_update: CompteComptableUpdate,
    db: Session = Depends(get_db)
):
    """Met à jour un compte comptable"""
    service = ComptabiliteService(db)
    try:
        return await service.update_compte(compte_id, compte_update.dict(exclude_unset=True))
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

# Endpoints pour les écritures comptables
@router.post("/ecritures", response_model=EcritureComptableResponse)
async def create_ecriture(
    ecriture: EcritureComptableCreate,
    db: Session = Depends(get_db)
):
    """Crée une nouvelle écriture comptable"""
    service = ComptabiliteService(db)
    try:
        return await service.create_ecriture(ecriture.dict())
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.get("/ecritures", response_model=List[EcritureComptableResponse])
async def list_ecritures(
    compte_id: Optional[UUID] = None,
    journal_id: Optional[UUID] = None,
    date_debut: Optional[date] = None,
    date_fin: Optional[date] = None,
    statut: Optional[str] = None,
    db: Session = Depends(get_db)
):
    """Liste les écritures comptables avec filtres"""
    service = ComptabiliteService(db)
    return await service.get_ecritures(
        compte_id=compte_id,
        journal_id=journal_id,
        date_debut=date_debut,
        date_fin=date_fin,
        statut=statut
    )

@router.post("/ecritures/{ecriture_id}/valider")
async def valider_ecriture(
    ecriture_id: UUID,
    validee_par_id: UUID,
    db: Session = Depends(get_db)
):
    """Valide une écriture comptable"""
    service = ComptabiliteService(db)
    try:
        return await service.valider_ecriture(ecriture_id, validee_par_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

# Endpoints pour les journaux comptables
@router.post("/journaux", response_model=JournalComptableResponse)
async def create_journal(
    journal: JournalComptableCreate,
    db: Session = Depends(get_db)
):
    """Crée un nouveau journal comptable"""
    service = ComptabiliteService(db)
    return await service.create_journal(journal.dict())

@router.get("/journaux", response_model=List[JournalComptableResponse])
async def list_journaux(
    type_journal: Optional[str] = None,
    actif: Optional[bool] = None,
    db: Session = Depends(get_db)
):
    """Liste les journaux comptables"""
    service = ComptabiliteService(db)
    return await service.get_journaux(type_journal=type_journal, actif=actif)

# Endpoints pour les exercices comptables
@router.post("/exercices", response_model=ExerciceComptableResponse)
async def create_exercice(
    exercice: ExerciceComptableCreate,
    db: Session = Depends(get_db)
):
    """Crée un nouvel exercice comptable"""
    service = ComptabiliteService(db)
    try:
        return await service.create_exercice(exercice.dict())
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/exercices/{annee}/cloturer")
async def cloturer_exercice(
    annee: str,
    cloture_par_id: UUID,
    db: Session = Depends(get_db)
):
    """Clôture un exercice comptable"""
    service = ComptabiliteService(db)
    try:
        return await service.cloturer_exercice(annee, cloture_par_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

# Endpoints pour les rapports comptables
@router.get("/grand-livre", response_model=List[LigneGrandLivre])
async def get_grand_livre(
    compte_id: Optional[UUID] = None,
    date_debut: Optional[date] = None,
    date_fin: Optional[date] = None,
    db: Session = Depends(get_db)
):
    """Génère le grand livre"""
    service = ComptabiliteService(db)
    return await service.get_grand_livre(
        compte_id=compte_id,
        date_debut=date_debut,
        date_fin=date_fin
    )

@router.get("/balance", response_model=List[CompteBalance])
async def get_balance(
    date_debut: Optional[date] = None,
    date_fin: Optional[date] = None,
    db: Session = Depends(get_db)
):
    """Génère la balance générale"""
    service = ComptabiliteService(db)
    return await service.get_balance_generale(
        date_debut=date_debut,
        date_fin=date_fin
    )

@router.get("/bilan", response_model=BilanResponse)
async def get_bilan(
    date_fin: date = Query(..., description="Date de fin pour le bilan"),
    db: Session = Depends(get_db)
):
    """Génère le bilan comptable"""
    service = ComptabiliteService(db)
    return await service.get_bilan(date_fin=date_fin)

@router.get("/compte-resultat", response_model=CompteResultatResponse)
async def get_compte_resultat(
    date_debut: date = Query(..., description="Date de début de l'exercice"),
    date_fin: date = Query(..., description="Date de fin de l'exercice"),
    db: Session = Depends(get_db)
):
    """Génère le compte de résultat"""
    service = ComptabiliteService(db)
    return await service.get_compte_resultat(
        date_debut=date_debut,
        date_fin=date_fin
    )
